clio-server/src/main/scala/org/broadinstitute/clio/server/service/IndexService.scala

```diff
@@ -67,4 +78,133 @@ abstract class IndexService[CI <: ClioIndex](
       elasticsearchIndex
     )
   }
+  private val keysToDrop =
+    Set(
+      ElasticsearchIndex.UpsertIdElasticsearchName,
+      ElasticsearchIndex.EntityIdElasticsearchName,
+    )
+
+  def validateUpsert(
+    indexKey: clioIndex.KeyType,
+    metadata: clioIndex.MetadataType,
+    force: Boolean
+  )(
+    implicit materializer: Materializer
+  ): Either[Exception, clioIndex.MetadataType] = {
+
+    // drop housekeeping and key fields (evaluate only metadata)
+    val dropKeys = keysToDrop ++ indexKey.getClass.getDeclaredFields
+      .map(field => ElasticsearchUtil.toElasticsearchName(field.getName))
+
+    val storedDocsForKey = Await
+      .result(
+        searchService
+          .simpleStringQueryMetadata(indexKey.asJson.pretty(Printer.noSpaces))(
+            elasticsearchIndex
+          )
+          .runWith(Sink.seq),
+        1.second
+      )
+      .map(
+        json =>
+          json.mapObject(_.filterKeys({ key =>
```

Comment:
It seems like this work (computing `dropKeys` and this filtering) can be done after the checks on line 115 and line 123. Doing it afterwords would make it clearer that the filtering done here doesn't affect the top level collection, just the json contents of the collection. (If that is what it is doing, it's a little hard to tell)